### DISCUSSION TITLE
Add permissions to conversation resolution

### DISCUSSION THREAD

**Discussion by jakeleventhal** [community/community#12961] [Pull Requests]
There should be some sort of permission system on conversation resolution. For instance, there should be an option to not allow the author of a PR to resolve conversations.

Example/use case:
John makes a PR
Bob leaves feedback in a conversation "you should fix this line"
John resolves the conversation without addressing the feedback
Bob forgets about the conversation since it is one of many conversations on the PR

If John was not able to resolve the conversation (only the PR reviewers can resolve) then this can't happen

**Comment by PavelA85** [comment]
agree, this would make life much easier.

**Comment by jfullstackdev** [comment]
but the owner of the repo still has the full rights whether to address the PR or not, 

if that is in a company, then it's their internal communication that will matter

**Reply by cferreira-fuze** [reply]
Which is why devs usually aren't the admins of the repo.

**Comment by cferreira-fuze** [comment]
I also agree that such feature would provide better PR handling. But this should only be valid for comments added in request for revision. Comments added only in the spirit of comments, should not be included in this restriction.

**Comment by cgaldiolo-splashlight** [comment]
Sometime while reviewing a PR I start a conversation to address a problem or a doubt.
If I start a conversation I would like to be the only one to have the permission to resolve that conversation (not the PR author and not another reviewer).
Many times my conversations are being resolved by someone else and I have to click "Show Resolved" just to check if it was closed 
 by me or someone else and why

**Reply by jakeleventhal** [reply]
For this reason, perhaps this feature can be implemented as a "Mark needs work" button that can be applied to any comment. essentially a task feature. The button could create 3 buttons (only one selectable at a time).

Needs work (selected)
Addressed
Closed

If the needs work button was never clicked, the three buttons never appear. If it is clicked, needs work is selected.

The reviewers can click each option. The PR author can only click needs work and addressed.

**Comment by DanJBower** [comment]
This sort of feature would be really helpful for my organisation. For our review process we'd like it so that all comments that might need a problem addressing should only be resolvable by the author of the comment or by the review moderator in times of deadlock between the comment author and the review owner. This alongside requiring all comments to be resolved before the PR can be complete would be perfect

**Comment by cobaltplay** [comment]
I would really like this feature. tracking down my conversations to see if they were answered is a pain

**Reply by TimCTP** [reply]
It's incredibly clunky, and painful. And the UI even hides some conversations in the middle. When you have a dev on the team that does some changes, ignores others, self-resolves etc, it's 3x as much work to figure out what's going on, and makes one consider getting a food truck.

**Comment by jeniabrook** [comment]
Is there any news regarding this awesome feature?

**Comment by jakeleventhal** [comment]
This is one of those features worth applying to work at GitHub, implementing, then quitting lol

**Comment by kdurando** [comment]
I completely agree that there's a need. I'm just about to unresolve 175 conversations to verify the implementation of what has been done.

**Comment by willsully** [comment]
Definitely in favour of this - but obviously it should be a configurable option to be tailored to each organisation/individual's needs.

I'd like the ability to lock it down so that only the person who raised the comment is able to mark it as resolved - that's the process we follow but unfortunately there's no way to enforce this. Also an admin override of some kind would probably be necessary.

**Comment by SaulBurgos** [comment]
I need this feature. Big companies with different timezone is vert handy

**Comment by jagerman** [comment]
This is definitely needed.

Additionally the "Require conversation resolution before merging" branch protection rule on a repository becomes fairly useless without comment resolution permissions because someone can trivially just resolve all conversations to bypass the protection rule.

**Reply by mroy-seedbox** [reply]
This 👆

**Comment by hohwille** [comment]
In general the permissions of github are way too corsare grained.
I do not want to give write permissions to my repository to (all) developers.
However, with triage permission developers cannot resolve conversations that they started if the PR is not created by them (in the initial example Bob does not have the permission to resolve his own review comment what IMHO does not make any sense to me).
Maybe you also want to allow some developers to run a workflow manually, etc. but you do not want to grant Admin permissions...
Even if you pay a silver or gold subscription you still need github enterprise to make use of fine-grained permissions.
That does not really make sense to me.

**Comment by amatiash** [comment]
I feel a need to restrict a person who made a PR to resolve your comments as a reviewer.
Recently had a case of a developer that closed my comment without actually making the fix.

**Reply by cferreira-fuze** [reply]
Unfortunately, dysfunctional teams are a reality and it's precisely because of such reality that these features are needed.

**Reply by mroy-seedbox** [reply]
Even it they resolve it themselves _after_ making the correct fix, it can be hell for reviewers to track when there are dozens of comments (or more).

I approach resolving conversations like checking off items on a list (_but it does not work if someone else starts checking off items for me, even if they did it_).

**Comment by pristinesource** [comment]
It would be great if we had a permission/setting to only allow the reviewer to resolve conversations on a pull request when requesting changes.  We just recently ran into a situation where this would have been very beneficial.

**Comment by TimCTP** [comment]
For starters (and I mean bare minimum, I strongly agree with the suggestion here), it would be great if the UI showed Resolved - username here, instead of just this:

![image](https://github.com/user-attachments/assets/7a424f40-684e-47c6-b791-5edc4cb0efd7)

Then you could at least tell the situation at a glance without having to expand each one (i.e. identify did the reviewer resolve, or the PR author).

**Reply by mroy-seedbox** [reply]
Oh yeah, that alone would be a huge improvement! 👌 

Then we could un-resolve and re-resolve ourselves if it is indeed fixed. 😅

**Reply by mroy-seedbox** [reply]
And the information is already available!

It just needs to be displayed differently (without having to click "Show resolved" to see it).

![image](https://github.com/user-attachments/assets/53244fdd-dd9e-4728-b15d-adbd2e030113)

**Comment by mroy-seedbox** [comment]
I am a very thorough reviewer, and it is hell to keep track of things if other people resolve my own comments.

And it's not like we can expect the problem to go away just by saying "please do not resolve my comments".

I'm tired of clicking "show resolved". I am not loving this experience _at all_. But it would be a lovely experience if I were the only one able to resolve my own comments.

Please fix this.

**Comment by mroy-seedbox** [comment]
Imagine a marriage like this:

> _"Honey, we need to talk."_

> _"Resolve conversation."_

> _"What?!???"_

**Reply by mroy-seedbox** [reply]
And if that doesn't work, there's aways "[sudo resolve conversation](https://xkcd.com/149/)". 😆 

![image](https://github.com/user-attachments/assets/05f73c04-9883-4f31-aee7-aee0ac938bbb)

**Comment by mroy-seedbox** [comment]
Right now, I have to track resolution by putting a 👍 emoji reaction at the end of the conversation, _**because no one else can put one for me**_. And even if they also put their own 👍 , it shows up differently (not in blue).

Sometimes it looks weird because I'm putting a 👍 on my own comment, but 🤷‍♂️.

_But "Require conversation resolution before merging" is still broken since people can self-resolve all issues_.

![image](https://github.com/user-attachments/assets/76255a90-c4f9-47ea-8fa5-b294a2efacf7)

**Comment by mroy-seedbox** [comment]
Alright, I got tired enough of this that I created a little bookmarklet to handle it. It's not perfect, but it helps a lot.

Basically how it works:
- Change the `username` to your own.
- Copy-paste the entire code as the URL of a bookmark in your bookmark bar
  - You can look at it, it's not doing anything dangerous.
- Click it while on the "Conversation" tab of a PR.

When clicked:
- It will expand all of the conversations that are not currently expanded.
- It will remove from the page/DOM those that are expanded and resolved by `username` (yourself).
  - And leave any which are unresolved or resolved by someone else.

So you might need to click it twice in order to have the full effect: first to expand, and then again to delete (because we cannot know who resolved a conversation without expanding it first).

If other people resolve your comments, simply unresolve them and then resolve them yourself.

_But of course, "Require conversation resolution before merging" is still broken since people can self-resolve all issues._

```js
javascript:(() => {
  const username = "mroy-seedbox";
  document.querySelectorAll('div.js-timeline-item:has(details.js-resolvable-timeline-thread-container > summary)').forEach(e => {
    if (e.querySelector('details').getAttribute('open') == null) {
      e.querySelector('details span.Details-content--closed').click();
    } else {
      const resolver = e.querySelector('details form.js-resolvable-timeline-thread-form > div > strong')?.innerText;
      if (resolver == username) e.remove();
    }
  });
})();
```

**Comment by zubin** [comment]
This is a complicated problem because there are multiple perspectives:
- PR authors want to mark conversations as resolved so they can see what's left to address. 
- PR reviewers want to revisit conversations they've raised to see whether they've been ~resolved~ addressed. 

Therefore, this behaviour may make sense:

- When a PR author marks a conversation as resolved, it's only resolved from their perspective. 
  - To everyone else it says "Marked as resolved by author" but it remains unfolded to other reviewers. 
- Reviewers can choose "Agreed: conversation is resolved", which will fold the conversation only for them.
  - Otherwise they can comment and optionally unresolve the conversation (for the author). 

A grey area is where to apply the non-author logic:
1. to anyone who commented in the conversation
2. to whoever initiated the conversation
3. to all reviewers
4. to everyone except the author

Personally, I prefer option 1. But this leads to another question: unless option 4, what should everyone else see? Probably same as author (ie resolved).

**Reply by TimCTP** [reply]
Not bad ideas. I hadn't thought about the "PR author tracks things done by using 'resolved'" angle (since I don't do that). Perhaps one "resolved" state isn't enough, like "done/not doing" (author picks) and "resolved" (reviewer or admin picks)?

**Reply by mroy-seedbox** [reply]
Good point! The behavior you propose makes sense. 👍 

One way to make it work for everyone could be for the system to remember the "Hide/show resolved" status for each conversation, _for each user_. And by default it should remain open (even if resolved) until a user manually hides it. And that could be a behavior toggle for users who find that annoying ("Show resolved conversation until hidden": true|false), which should be false/unchecked by default (to preserve the current behavior).

That way, "Hide resolved" would become a way for each user to mark a conversation as completed _for them_.

**Reply by mroy-seedbox** [reply]
Actually, see @jakeleventhal 's comment below. It's an even better solution.

**Reply by TimCTP** [reply]
^ YES. Ooh la la.

**Comment by jakeleventhal** [comment]
When I worked at SpaceX, the way we had issue resolution was like this:

Every comment had a button group with 3 options that acted like radio buttons
1. Needs work
2. Addressed
3. Resolved

PR authors could only select 1 or 2
PR reviewers could select 1, 2, or 3

**Reply by mroy-seedbox** [reply]
👌 

Addressed vs Resolved is the perfect way to solve this issue with "resolved by PR author" vs "resolved by PR reviewer".

And if authors can only mark as "Addressed" instead of "Resolved", then they would no longer be able to bypass "Require conversation resolution before merging" by marking everything as resolved themselves.

**Reply by zubin** [reply]
Agreed: "addressed vs resolved" is simple and intuitive. I like!

**Comment by thomas-j-wtfoxtrot** [comment]
I also like "addressed vs resolved".

I would add a configuration option who can resolve (using above list):
* to whoever initiated the conversation
* to anyone who commented in the conversation
* to all reviewers
* to everyone except the author
* to everyone

I would support the 5th option to provide the present behavior as well, for those who want.

**Comment by XedinUnknown** [comment]
I think this can be much simpler: the new option could prevent the resolution of threads opened by other people, unless the user is the owner/maintainer of the repo, in which case they can force-resolve (just like they do with merges).

My premise is simple: when user `A` opens a thread, they own the acceptance criteria of that thread (whether to answer a question or to address a request), and only they can therefore validate whether this criteria is met. Admins or maintainers should be able to do whatever they want, but if they resolve someone else's thread - that should be a very explicit action, ensuring they are aware that they are breaking the rules. If the owner doesn't want those rules - he simply doesn't tick that box, and everything keeps working like it is today.

**Reply by zubin** [reply]
This is a simpler approach, and definitely an improvement, however it's less flexible since PR authors can't flag comments as being addressed—which is legitimate and useful IMO—unless they force-resolve them (which defeats the purpose).

**Reply by XedinUnknown** [reply]
Why does that defeat the purpose? A thread is open because there's something to address. Closing it = marking it as resolved. The person who opens it - should be the person to validate whether it is addressed. If they are suddenly inactive, the maintainer can decide that instead: otherwise, if the settings require all threads to be closed for merge, the maintainer won't be able to merge a PR in their own repo - just because someone won't close their own thread.

Or do you mean that, if a change is requested of me, I should be able to indicate that I've addressed it? That can already be done: I just reply in that thread with:

> ✅ Done in commit `123456`

**Only the author of the thread, and therefore the owner of the acceptance criteria, can determine whether their request has really been resolved** (besides the maintainers).

**Reply by jakeleventhal** [reply]
@XedinUnknown i dont think youre seeing the point. The feature is also opt-in

**Reply by XedinUnknown** [reply]
Could be. My point is that many of the suggestions here are quite complicated and opinionated. Maybe it will be less controversial and therefore quicker to accept and implement if it was simple and covered most cases. So I tried to propose a simpler approach, which aligns with many of the already existing features and usage patterns. Surely I must have missed something. But if you like the simplicity of my approach - feel free to extend it 🙏

**Reply by downsider** [reply]
This is exactly what we do in my team at work. We currently enforce it with the "baseball bat" policy. e.g. "If *you* resolve __my__ comment, I'll be paying you a visit". Having this as a full feature in Github would save countless hours tidying up HR issues 😄 

I feel there is a distinct advantage in keeping this simple, if only because it's easier to implement. Like @XedinUnknown says, marking a comment as addressed can be as simple as replying with "done". Github sends me emails whenever someone comments on my threads or pushes commits to a PR I've reviewed so I don't personally see the need for anything new here. Keeping it simple also allows for the feature to be built upon later, if the feedback is that more nuance is required.

**Comment by mroy-seedbox** [comment]
Just stumbled upon [this](https://mtlynch.io/code-review-love/) as another example:

<img width="635" height="578" alt="image" src="https://github.com/user-attachments/assets/7d720918-5bdc-49fc-bf3b-be7614a48e4a" />

**Comment by JamesMcDaniel04** [comment]
Agreed,  but the owner of the repo also has the ability to reject and add commentary about expectations

**Comment by gethari** [comment]
Wow its  

# 2026 

and still not happening

**Comment by xs2bharat** [comment]
Thinking to create GitHub action flow which can go over all conversations and get list of conversation which is resolved by author but GitHub feature would be nice to have to avoid this custom thing

**Comment by soham-suki** [comment]
bump! Would love this, please!

**Comment by JenChitty-Aarcomm** [comment]
Yeah, this looks like another essential feature that GitHub just doesn't care about adding. Either they've failed to leverage Copilot to accelerate their development after all, they've lost control of their backlog, or they don't care about their users anymore. Heck, they money is rolling in, so I guess from their perspective it aint broke so why fix it?

GitHub, please prove me wrong!

**Comment by Pramuspl** [comment]
+1. Would be of great help!

**Comment by JamesMcDaniel04** [comment]
❤️
